helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -159,12 +159,19 @@ private float getAssignmentNormalizedScore(AssignableNode node, AssignableReplic
           int statePriority1 = replica1.getStatePriority();
           int statePriority2 = replica2.getStatePriority();
           if (statePriority1 == statePriority2) {
-            // If state prioritizes are the same, compare the names.
-            if (resourceName1.equals(resourceName2)) {
-              return replica1.getPartitionName().compareTo(replica2.getPartitionName());
-            } else {
-              return resourceName1.compareTo(resourceName2);
-            }
+            // If state prioritizes are the same, try to randomize the replicas order. Otherwise,
+            // the same replicas might always be moved in the rebalance. This is because their
+            // rebalance calculating will always happen at the critical moment while the cluster is
+            // almost full.
+            //
+            // Note that to ensure the algorithm is deterministic with the same inputs, do not use
+            // Random functions here. Use hashcode based on the cluster topology information to get
+            // a controlled randomized order is good enough.
+            long replicaHash1 = Objects
```

Comment:
Is this hashing algorithm deterministic?

Reply:
It is java hashing based. Just a wrapper.